### PR TITLE
Ability to mock Flow APIs

### PR DESCRIPTION
**Pull request by aksharp**

### CODE REVIEW

**Comment by paololim** [comment]
Perhaps I miss some context here, but it looks like `api-mocker` is just a collection of enums for PFS/Inventory: https://github.com/flowcommerce/api-misc/blob/master/spec/api-mocker.json. 

This seems like very specific code that should have nothing to do with lib-play. Note that the updated methods are literally used everywhere - do we know if those actually work by testing? Also, why not extend the necessary controllers in only the relevant repos?

**Comment by aksharp** [comment]
Yes, there’s more context @paololim. api-mocker was initially created to mock individual external apis and has evolved since then. Maybe a demo would be helpful.

**Comment by aksharp** [comment]
Here are few use cases: 
1) Testing 
a) mock cross-service api calls when testing an individual service
b) mock external api calls (wrapped by Flow api) when testing in prod
c) Front-end can mock 422, 500 and other errors easily to test UI functionality
2) Development process
a) When new endpoint is up, it is automatically mockable for parallel development (UI or Cross-service)

**Comment by paololim** [comment]
Not really only the PFS-specific logic. I am just thinking that if we want to enable this for *any* of the APIs, then probably the answer isn't to create an `api-mocker` api, but to figure out how to use an attribute(s)/hints to tag resources as "mockable" or something along those lines. They way it looks set up now, do we really have to add to the enum?

**Comment by aksharp** [comment]
Enum is not used for lib-play and can be moved from api-mocker. I can explain the details and my thoughts on implementation beyond this PR tomorrow in person.

File: app/io/flow/play/controllers/FlowActionInvokeBlockHelper.scala [hunk 1]
```diff
@@ -12,11 +16,26 @@ trait FlowActionInvokeBlockHelper {
 
   def jwtSalt: String = config.requiredString("JWT_SALT")
 
+  val `X-Flow-Mock-Api` = "X-Flow-Mock-Api"
+  val `X-Flow-Mock-Api-Secret` = "X-Flow-Mock-Api-Secret"
+  def mockApiSecret: String = config.requiredString("MOCK_API_SECRET")
```
Comment (paololim) [inline app/io/flow/play/controllers/FlowActionInvokeBlockHelper.scala]: Every single service will likely fail startup by making this required

File: app/io/flow/play/controllers/FlowController.scala [hunk 2]
```diff
@@ -86,7 +86,9 @@ class AnonymousActionBuilderImpl[B](val parser: BodyParser[B], val config: Confi
       // Create an empty header here so at least requestId tracking can start
       AuthData.Anonymous.Empty
     }
-    block(new AnonymousRequest(ad, request))
+    mockApi(request.headers)
```
Comment (paololim) [inline app/io/flow/play/controllers/FlowController.scala]: Adding this here seems dangerous... Per my earlier comment - this is pretty much on the code path of every single resource operation we use at Flow.

File: app/io/flow/play/controllers/FlowActionInvokeBlockHelper.scala (hunk 1 above)
Comment (aksharp) [reply app/io/flow/play/controllers/FlowActionInvokeBlockHelper.scala]: Consuming services would only require MOCK_API_SECRET environment variable.

Consuming
- From Backend: add 2 requestHeaders on generated client:
1) X-Flow-Mock-Api-Response = stringify(MockApiResponse(httpStatusCode, body = Json.toJson(your type)), contentType = defaults to application/json))
2) X-Flow-Mock-Api-Secret = config.requiredString("MOCK_API_SECRET")

- From Frontend: same as above, except X-Flow-Mock-Api-Response is in JSON straight away

File: app/io/flow/play/controllers/FlowController.scala (hunk 2 above)
Comment (aksharp) [reply app/io/flow/play/controllers/FlowController.scala]: yes, the intent is to have every action to be mockable without extra effort

File: app/io/flow/play/controllers/FlowActionInvokeBlockHelper.scala (hunk 1 above)
Comment (aksharp) [reply app/io/flow/play/controllers/FlowActionInvokeBlockHelper.scala]: P.S. Not intended to be merged until each service is set up with MOCK_API_SECRET environment variable

**Comment by aksharp** [comment]
Found a better way to mock api (via client) Closing this PR

**Pull request closed without merging**